Sources/NIOTestUtils/EventCounterHandler.swift

```diff
@@ -0,0 +1,363 @@
+        }
+        guard self.connectCalls <= 1 else {
+            throw ValidityError("connect should be called no more than once")
+        }
+        guard self.closeCalls <= 1 else {
+            throw ValidityError("close should be called no more than once")
+        }
+    }
+
+    /// Return all event descriptions that have triggered.
+    ///
+    /// This is most useful in unit tests where you want to make sure only certain events have been triggered.
+    ///
+    /// - note: This API is thread-safe, you may call it from any thread. The results of this API may vary though if you
+    ///         call it whilst the `Channel` this `ChannelHandler` is in is still in use.
+    public func allTriggeredEvents() -> Set<String> {
+        var allEvents: [(String, Int)] = []
+
+        if self.channelRegisteredCalls != 0 {
+            allEvents.append(("channelRegistered", self.channelRegisteredCalls))
+        }
+
+        if channelUnregisteredCalls != 0 {
+            allEvents.append(("channelUnregistered", self.channelUnregisteredCalls))
+        }
+
+        if channelActiveCalls != 0 {
+            allEvents.append(("channelActive", self.channelActiveCalls))
+        }
+
+        if channelInactiveCalls != 0 {
+            allEvents.append(("channelInactive", self.channelInactiveCalls))
+        }
+
+        if channelReadCalls != 0 {
+            allEvents.append(("channelRead", self.channelReadCalls))
+        }
+
+        if channelReadCompleteCalls != 0 {
+            allEvents.append(("channelReadComplete", self.channelReadCompleteCalls))
+        }
+
+        if channelWritabilityChangedCalls != 0 {
+            allEvents.append(("channelWritabilityChanged", self.channelWritabilityChangedCalls))
+        }
+
+        if userInboundEventTriggeredCalls != 0 {
+            allEvents.append(("userInboundEventTriggered", self.userInboundEventTriggeredCalls))
+        }
+
+        if errorCaughtCalls != 0 {
+            allEvents.append(("errorCaught", self.errorCaughtCalls))
+        }
+
+        if registerCalls != 0 {
+            allEvents.append(("register", self.registerCalls))
+        }
+
+        if bindCalls != 0 {
+            allEvents.append(("bind", self.bindCalls))
+        }
+
+        if connectCalls != 0 {
+            allEvents.append(("connect", self.connectCalls))
+        }
+
+        if writeCalls != 0 {
+            allEvents.append(("write", self.writeCalls))
+        }
+
+        if flushCalls != 0 {
+            allEvents.append(("flush", self.flushCalls))
+        }
+
+        if readCalls != 0 {
+            allEvents.append(("read", self.readCalls))
+        }
+
+        if closeCalls != 0 {
+            allEvents.append(("close", self.closeCalls))
+        }
+
+        if triggerUserOutboundEventCalls != 0 {
+            allEvents.append(("triggerUserOutboundEvent", self.triggerUserOutboundEventCalls))
+        }
+
+        return Set(allEvents.map { $0.0 })
```

Comment:
This function is weird. Why does it create an intermediate array? Why does it load the atomic twice? Why does only the second load use the explicit `self`? Why does it perform the second load at all?

Reply:
oops, leftovers, will fix. Thank you!